### DISCUSSION TITLE
What is the Nautilus 'github-actions-secrets' ?

### DISCUSSION THREAD

**Discussion by da2ce7** [Nautilus-Cyberneering/github-actions-secrets#2] [General]
Is this the repository where workflows are run without permissions? hmmm?

**Comment by josecelano** [comment]
Hi @da2ce7 . No, it is not.

It's just a test repo to know more about the workflow lifecycle. I want to:

- Explain the problem of sharing secrets with third-party actions.
- Add some examples to understand how secrets are handled by GitHub.
- Propose some solutions or good practices.

It's a kind of article-repo I like to do like [this](https://github.com/Nautilus-Cyberneering/pygithub), but my goal is to write only one article in the README.

Sorry for not writing at least a basic README with these goals. I thought I could finish it yesterday but then I realized secrets are handled in a different way than what I thought. My main misconception was I thought secrets were accessible by default to actions (when the workflow is executed in a branch by a maintainer) and that seems not to be true. Anyway, I'm adding some examples to check it.

### NOTES

- GitHub uses contexts to pass information to the runners.
- There is one special context called "[secrets](https://docs.github.com/en/actions/learn-github-actions/contexts#secrets-context)".
- The secret context contains a special secret that GitHub adds by default: [GITHUB_TOKEN](https://docs.github.com/en/actions/security-guides/automatic-token-authentication).
- The `GITHUB_TOKEN` has its own rules about what permissions have by default depending on where you execute the workflow.
- The secret context is not always available. See [context availability](https://docs.github.com/en/actions/learn-github-actions/contexts#context-availability).

### IMPORTANT

- Actions do not have access to the secrets by default, even if the `GITHUB_TOKEN` in the workflow has full permissions. That was another misconception, I thought you can get the secrets via the API using the `GITHUB_TOKEN` but that's not true. The only way to pass a secret to an action is with an input or an env var.
-  For example, with `MegaLinter` we are passing one of the secrets `secrets.GITHUB_TOKEN`. With that secret you cannot get directly other secrets, what you can do (if you have "write" permissions) is to create a new workflow to get the secrets.

### TODO

- [ ] Write article.
- [X] [Example of docker action to show what the action gets from env vars](https://github.com/Nautilus-Cyberneering/github-actions-secrets/runs/6629060237?check_suite_focus=true#step:3:12).
- [ ] Example of TypeScript action to show what the action gets from env vars.
- [ ] I'm not sure if secrets are stored somewhere in the context of the runner and the action could read that file from the disk even if you do not pass the secret as an input or env var.
- [ ] If you set a job or workflow env var with a secret, I assume it would be also available to all actions.

### CONCLUSION

- You should be safe as long as you pass a `GITHUB_TOKEN` only with read permissions for contents.
- You can even set `CODEOWNERS` for workflows folders, so even if you pass a "write:content" token third-party actions cannot change workflows.
- We should be even more explicit with permissions and remove also "read" permissions if it's not needed.
